File: src/types/proposal.ts

import { FileText, Clock, Send, CheckCircle, XCircle, AlertTriangle, type LucideIcon } from "lucide-react";

export type ProposalStatus = 'draft' | 'sent' | 'accepted' | 'rejected' | 'expired' | 'pending_approval';

// Status label mapping for UI display
export const proposalStatusLabels: Record<ProposalStatus, string> = {
  draft: 'Taslak',
  pending_approval: 'Onay Bekliyor',
  sent: 'Gönderildi',
  accepted: 'Kabul Edildi',
  rejected: 'Reddedildi',
  expired: 'Süresi Dolmuş'
};

// Status icon mapping for UI display  
export const proposalStatusIcons: Record<ProposalStatus, LucideIcon> = {
  draft: FileText,
  pending_approval: Clock,
  sent: Send,
  accepted: CheckCircle,
  rejected: XCircle,
  expired: AlertTriangle
};

// Simplified status color mapping - using semantic colors
export const proposalStatusColors: Record<ProposalStatus, string> = {
  draft: 'bg-muted text-muted-foreground',
  pending_approval: 'bg-yellow-100 text-yellow-800 dark:bg-yellow-900/20 dark:text-yellow-400',
  sent: 'bg-blue-100 text-blue-800 dark:bg-blue-900/20 dark:text-blue-400',
  accepted: 'bg-green-100 text-green-800 dark:bg-green-900/20 dark:text-green-400',
  rejected: 'bg-red-100 text-red-800 dark:bg-red-900/20 dark:text-red-400',
  expired: 'bg-orange-100 text-orange-800 dark:bg-orange-900/20 dark:text-orange-400'
};

export interface ProposalItem {
  id: string;
  product_id?: string;
  name: string;
  description?: string;
  quantity: number;
  unit: string; // Unit of measurement (adet, saat, gün, etc.)
  unit_price: number;
  tax_rate?: number;
  discount_rate?: number;
  total_price: number;
  currency?: string;
  group?: string; // Group information (e.g., product, service)
  stock_status?: 'in_stock' | 'low_stock' | 'out_of_stock';
  original_currency?: string; // Original currency of the product
  original_price?: number; // Original price of the product
}

export interface ProposalAttachment {
  id: string;
  name: string;
  url: string;
  type: string;
  size: number;
  created_at: string;
}

// Customer and Employee interfaces to support relations
export interface Customer {
  id: string;
  name: string;
  email?: string;
  company?: string;
  phone?: string;
  mobile_phone?: string;
  office_phone?: string;
  address?: string;
  tax_number?: string;
  tax_office?: string;
}

export interface Employee {
  id: string;
  first_name: string;
  last_name: string;
  email?: string;
  phone?: string;
  position?: string;
  department?: string;
  avatar_url?: string;
}

export interface Proposal extends Record<string, unknown> {
  id: string;
  number: string;
  title: string;
  description?: string;
  customer_id?: string;
  customer_name?: string;
  customer_email?: string;
  employee_id?: string;
  employee_name?: string;
  opportunity_id?: string;
  opportunity_title?: string;
  valid_until?: string;
  payment_terms?: string;
  payment_method?: string;
  delivery_terms?: string;
  delivery_method?: string;
  warranty_terms?: string;
  price_terms?: string;
  other_terms?: string;
  notes?: string;
  terms?: string;
  status: ProposalStatus;
  created_at: string;
  updated_at: string;
  total_amount: number;
  currency?: string; // Currency
  items?: ProposalItem[];
  attachments?: ProposalAttachment[];
  
  // Additional fields used in various components
  proposal_number?: string; // Backward compatibility field
  total_value?: number; // Backward compatibility field for total_amount
  internal_notes?: string;
  discounts?: number;
  additional_charges?: number;
  
  // Relations
  customer?: Customer;
  employee?: Employee;
}

export interface ProposalFormValues {
  title: string;
  description?: string;
  customerId?: string;
  employeeId?: string;
  opportunityId?: string;
  validUntil?: Date;
  paymentTerms?: string;
  deliveryTerms?: string;
  notes?: string;
  status: ProposalStatus;
  items?: ProposalItem[];
  currency?: string; // Currency
}

File: lib/models/proposal.dart
class Proposal {
  final String id;
  final String number;
  final String title;
  final String? description;
  final String? customerId;
  final String? opportunityId;
  final String status;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? validUntil;
  final double totalAmount;
  final String currency;
  final String? terms;
  final String? notes;
  final String? employeeId;
  final List<dynamic>? items;
  final List<dynamic>? attachments;
  final String? paymentTerms;
  final String? deliveryTerms;
  final String? warrantyTerms;
  final String? priceTerms;
  final String? otherTerms;
  final List<String>? selectedPaymentTerms;
  final List<String>? selectedDeliveryTerms;
  final List<String>? selectedWarrantyTerms;
  final List<String>? selectedPricingTerms;
  final List<String>? selectedOtherTerms;
  final String? companyId;
  final String? subject;
  final double exchangeRate;
  final DateTime? offerDate;
  final List<dynamic>? history;
  final String? parentProposalId;
  final int? revisionNumber;

  Proposal({
    required this.id,
    required this.number,
    required this.title,
    this.description,
    this.customerId,
    this.opportunityId,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.validUntil,
    required this.totalAmount,
    required this.currency,
    this.terms,
    this.notes,
    this.employeeId,
    this.items,
    this.attachments,
    this.paymentTerms,
    this.deliveryTerms,
    this.warrantyTerms,
    this.priceTerms,
    this.otherTerms,
    this.selectedPaymentTerms,
    this.selectedDeliveryTerms,
    this.selectedWarrantyTerms,
    this.selectedPricingTerms,
    this.selectedOtherTerms,
    this.companyId,
    this.subject,
    required this.exchangeRate,
    this.offerDate,
    this.history,
    this.parentProposalId,
    this.revisionNumber,
  });

  factory Proposal.fromJson(Map<String, dynamic> json) {
    return Proposal(
      id: json['id'] ?? '',
      number: json['number'] ?? '',
      title: json['title'] ?? '',
      description: json['description'],
      customerId: json['customer_id'],
      opportunityId: json['opportunity_id'],
      status: json['status'] ?? 'draft',
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
      validUntil: json['valid_until'] != null
          ? DateTime.parse(json['valid_until'])
          : null,
      totalAmount: (json['total_amount'] ?? 0).toDouble(),
      currency: json['currency'] ?? 'TRY',
      terms: json['terms'],
      notes: json['notes'],
      employeeId: json['employee_id'],
      items: json['items'] != null 
          ? (json['items'] is List 
              ? List<dynamic>.from(json['items']) 
              : null)
          : null,
      attachments: json['attachments'] != null
          ? (json['attachments'] is List
              ? List<dynamic>.from(json['attachments'])
              : null)
          : null,
      paymentTerms: json['payment_terms'],
      deliveryTerms: json['delivery_terms'],
      warrantyTerms: json['warranty_terms'],
      priceTerms: json['price_terms'],
      otherTerms: json['other_terms'],
      selectedPaymentTerms: json['selected_payment_terms'] != null
          ? (json['selected_payment_terms'] is List
              ? List<String>.from(json['selected_payment_terms'])
              : null)
          : null,
      selectedDeliveryTerms: json['selected_delivery_terms'] != null
          ? (json['selected_delivery_terms'] is List
              ? List<String>.from(json['selected_delivery_terms'])
              : null)
          : null,
      selectedWarrantyTerms: json['selected_warranty_terms'] != null
          ? (json['selected_warranty_terms'] is List
              ? List<String>.from(json['selected_warranty_terms'])
              : null)
          : null,
      selectedPricingTerms: json['selected_pricing_terms'] != null
          ? (json['selected_pricing_terms'] is List
              ? List<String>.from(json['selected_pricing_terms'])
              : null)
          : null,
      selectedOtherTerms: json['selected_other_terms'] != null
          ? (json['selected_other_terms'] is List
              ? List<String>.from(json['selected_other_terms'])
              : null)
          : null,
      companyId: json['company_id'],
      subject: json['subject'],
      exchangeRate: (json['exchange_rate'] ?? 1).toDouble(),
      offerDate: json['offer_date'] != null
          ? DateTime.parse(json['offer_date'])
          : null,
      history: json['history'] != null
          ? (json['history'] is List
              ? List<dynamic>.from(json['history'])
              : null)
          : null,
      parentProposalId: json['parent_proposal_id'],
      revisionNumber: json['revision_number'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'number': number,
      'title': title,
      'description': description,
      'customer_id': customerId,
      'opportunity_id': opportunityId,
      'status': status,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'valid_until': validUntil?.toIso8601String(),
      'total_amount': totalAmount,
      'currency': currency,
      'terms': terms,
      'notes': notes,
      'employee_id': employeeId,
      'items': items,
      'attachments': attachments,
      'payment_terms': paymentTerms,
      'delivery_terms': deliveryTerms,
      'warranty_terms': warrantyTerms,
      'price_terms': priceTerms,
      'other_terms': otherTerms,
      'selected_payment_terms': selectedPaymentTerms,
      'selected_delivery_terms': selectedDeliveryTerms,
      'selected_warranty_terms': selectedWarrantyTerms,
      'selected_pricing_terms': selectedPricingTerms,
      'selected_other_terms': selectedOtherTerms,
      'company_id': companyId,
      'subject': subject,
      'exchange_rate': exchangeRate,
      'offer_date': offerDate?.toIso8601String(),
      'history': history,
      'parent_proposal_id': parentProposalId,
      'revision_number': revisionNumber,
    };
  }
}

